### PR TITLE
Early stop

### PR DESCRIPTION
**Pull request by shrit**
This pull request updates the early stop callback to stop the training when the loss stops decreasing on the validation loss rather than the training loss.

### CODE REVIEW

File: mnist_batch_norm/mnist_batch_norm.cpp [hunk 1]
```diff
@@ -144,8 +144,16 @@ int main()
...
-              ens::EarlyStopAtMinLoss());
-
+              ens::EarlyStopAtMinLoss<arma::mat>(
+                  [&](const arma::mat& /* param */)
+                  {
+                    double validationLoss =
+                    model.Evaluate(validX, validY);
+                    std::cout << "Validation loss: "
+                              << validationLoss
+                              << std::endl;
```
Comment (zoq) [inline mnist_batch_norm/mnist_batch_norm.cpp:154]:
```suggestion
                        model.Evaluate(validX, validY);
                        std::cout << "Validation loss: "
                                  << validationLoss
                                  << std::endl;
```

Hopefully the idendation is correct (tab twice - 4 spaces).

File: mnist_cnn/mnist_cnn.cpp [hunk 2]
```diff
@@ -179,7 +179,16 @@ int main()
               ens::PrintLoss(),
               ens::ProgressBar(),
               // Stop the training using Early Stop at min loss.
-              ens::EarlyStopAtMinLoss());
+              ens::EarlyStopAtMinLoss<arma::mat>(
+                  [&](const arma::mat& /* param */)
+                  {
+                    double validationLoss =
+                    model.Evaluate(validX, validY);
```
Comment (zoq) [inline mnist_cnn/mnist_cnn.cpp:186]: See comment above.

File: mnist_simple/mnist_simple.cpp [hunk 3]
```diff
@@ -130,7 +130,16 @@ int main()
               ens::PrintLoss(),
               ens::ProgressBar(),
               // Stop the training using Early Stop at min loss.
-              ens::EarlyStopAtMinLoss());
+              ens::EarlyStopAtMinLoss<arma::mat>(
+                  [&](const arma::mat& /* param */)
+                  {
+                    double validationLoss =
+                    model.Evaluate(validX, validY);
```
Comment (zoq) [inline mnist_simple/mnist_simple.cpp:137]: See comment above.

File: mnist_batch_norm/mnist_batch_norm.cpp (hunk 1 above)
Comment (shrit) [reply mnist_batch_norm/mnist_batch_norm.cpp]: I think the 4 spaces are special lambda right?

Comment (shrit) [reply mnist_batch_norm/mnist_batch_norm.cpp]: I see because I have just cut the lines
